### DISCUSSION TITLE
Rework default value functions & commands

### DISCUSSION THREAD

**Discussion by Fusezion** [SkriptLang/Skript#5239] [Suggestions]
After some serious thoughts about how `<type=value>` in commands and functions were brought up about relating to `<type=value>` not requiring an optional surrounding `[]` placement, I went to talk with ChatGPT while thinking about it and ended up having conversations in #General in SkUnity discord.

The term we commonly use for `<type=value>` is DEFAULT VALUE, but this isn't the case for skript as we don't read it as a default value but rather a placeholder value that if the inputted argument is invalid, it returns a `null` value instead of the `value` input.

How to go About Changing
---

If anything, the <type=value> should require the `<=>` to be `[<=>]` while this change is relatively hurtful and causes breaking changes
this is the easiest way to define it as a nullable argument explicitly and makes a default value easier to understand

Should `<=>` be changed to work more as a default value? Similar to `{_var} ? "Empty"` if the value input is null, we set it to the secondary output, "Empty." Like a placeholder, if an invalid input is included, it will usually return the invalid input as it's technically been defined.

By changing the functionality of the placeholder, we create more of a definitive answer to how optionals and default values work while introducing correct terminology to this part of skript.

for those questioning why I mentioned terminology here*
> When I first brought up the mention of "placeholders" and "default values" a counter-argument was thrown related to the fact most people view "placeholder" as something like `%player's balance%` which is correct in the sense of plugins and PAPI. People who aren't native to skript or that term also think like this. However, I disagree with this argument as skript shouldn't be limited to what someone new to skript understands. If you're learning skript, you need to learn new terms you aren't familiar with along with correct terms to things you know.

Should we keep the placeholder style of `<type=placeholder>`?
> I'm strongly against us keeping this style in Skript. Skript has no use for placeholders, unlike where default values provide much more help across everything.

Optional:
> While we could keep the `<=>` and not enforce `[]` it would mean we have to force `<=>` to return an error when ran and no value >> is defined. This is so we can still label it a non-nullable argument, but if null after it is defined, we default to its default value.

How about functions?
---

Functions currently use a similar method to commands; instead of requiring `<>`. They act like it's defined without it existing. However, this led to functions never adding an optional argument selector, and in place, people used placeholder inputs using `=%placeholder%`.

This had become far too common across files and created many issues when invalid inputs got defined for both required and "optional" arguments.

Skript doesn't check if they're null and instead says "LGTM" and continues forwards. Skript should become more strict with this and introduce an optional tag like `example:?string`, where `?` is being used as the equivalent of `-` in skript's syntax for nullable. While doing this, we can force NonNullable arguments `example:string` to stop if the input is null. While still allowing us to do `example:?string="Hello, World!"`

**Related Issue** #5238
**Related Conversations** https://discord.com/channels/135877399391764480/135877399391764480/1050257575393898617

**Comment by sovdeeth** [comment]
I'm just going to mirror what I said in #5238:
> Personally, I'm of the opinion it shouldn't make anything optional, and optional arguments should only be those surrounded by `[]`. The behavior of the argument becoming optional is un-intuitive, as normally the only optional arguments in commands are surrounded by `[]`. Since this argument isn't, it appears required but actually isn't. Enforcing `[]` for all optional commands would also mean commands would more closely follow the generally accepted convention for Minecraft commands:
![image](https://user-images.githubusercontent.com/10354869/206355000-512310ef-6358-484c-bb13-9b16bf81d174.png)

For functions, I think what was suggested in #2993 is a good idea, more readable and more english-like than something like `?`. I don't have any strong opinions on how functions work though, I'm more concerned with the command arguments.

**Comment by Fusezion** [comment]
Updated my original message to become a bit more clear and also give more input to functions and make it more readable
